sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -2189,3 +2189,293 @@ case class ArrayRemove(left: Expression, right: Expression)
+            }
+          })
+          // store elements into array
+          val resultArray = new Array[Int](hsSize.size)
+          val hs = new OpenHashSet[Int]
+          var pos = 0
+          Seq(array1, array2).foreach(array => {
+            var i = 0
+            while (i < array.numElements () ) {
+              val elem = array.getInt (i)
+              if (!hs.contains (elem) ) {
+                resultArray (pos) = elem
+                hs.add (elem)
+                pos += 1
+              }
+              i += 1
+            }
+          })
+          if (ArraySetLike.useGenericArrayData(IntegerType.defaultSize, resultArray.length)) {
+            new GenericArrayData(resultArray)
+          } else {
+            UnsafeArrayData.fromPrimitiveArray(resultArray)
+          }
+        case LongType =>
+          // avoid boxing of primitive long array elements
+          // calculate result array size
+          val hsSize = new OpenHashSet[Long]
+          Seq(array1, array2).foreach(array => {
+            var i = 0
+            while (i < array.numElements()) {
+              if (hsSize.size > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+                ArraySetLike.throwUnionLengthOverflowException(hsSize.size)
+              }
+              hsSize.add(array.getLong(i))
+              i += 1
+            }
+          })
+          // store elements into array
+          val resultArray = new Array[Long](hsSize.size)
+          val hs = new OpenHashSet[Long]
+          var pos = 0
+          Seq(array1, array2).foreach(array => {
+            var i = 0
+            while (i < array.numElements()) {
+              val elem = array.getLong(i)
+              if (!hs.contains(elem)) {
+                resultArray(pos) = elem
+                hs.add(elem)
+                pos += 1
+              }
+              i += 1
+            }
+          })
+          if (ArraySetLike.useGenericArrayData(LongType.defaultSize, resultArray.length)) {
+            new GenericArrayData(resultArray)
+          } else {
+            UnsafeArrayData.fromPrimitiveArray(resultArray)
+          }
+        case _ =>
+          val arrayBuffer = new scala.collection.mutable.ArrayBuffer[Any]
+          val hs = new OpenHashSet[Any]
+          Seq(array1, array2).foreach(array => {
+            var i = 0
+            while (i < array.numElements()) {
+              val elem = array.get(i, elementType)
+              if (!hs.contains(elem)) {
+                if (arrayBuffer.size > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+                  ArraySetLike.throwUnionLengthOverflowException(arrayBuffer.size)
+                }
+                arrayBuffer += elem
+                hs.add(elem)
+              }
+              i += 1
+            }
+          })
+          new GenericArrayData(arrayBuffer)
+      }
+    } else {
+      ArraySetLike.evalUnionContainsNull(array1, array2, elementType,
+        if (elementTypeSupportEquals) null else ordering)
```

Comment:
If we use `ordering` for `elementTypeSupportEquals`, we have to use a loop instead of a hash function in `evalUnionContainsNull`. I think that to use a loop will lead to performance degradation compared to use a hash.
This is why we do not try to use `ordering` as possible.